src/orquestra/sdk/_base/_storage/orqdantic.py

```diff
@@ -0,0 +1,123 @@
+################################################################################
+# © Copyright 2024 Zapata Computing Inc.
+################################################################################
+
+"""Compatibility layer for pydantic v1 / v2 compatibility."""
+
+from copy import deepcopy
+from typing import Annotated, Any, Optional
+
+import pydantic
+
+PYDANTICV1 = pydantic.__version__.startswith("1.")
+
+if PYDANTICV1:
+    from pydantic.generics import GenericModel
+
+    class BaseModel(GenericModel):
+        @classmethod
+        def model_validate(cls, *args, **kwargs):
+            return super(BaseModel, cls).parse_obj(*args, **kwargs)
+
+        @classmethod
+        def model_validate_json(cls, *args, **kwargs):
+            return super(BaseModel, cls).parse_raw(*args, **kwargs)
+
+        def model_dump(self, *args, **kwargs):
+            return super().dict(*args, **kwargs)
+
+        def model_dump_json(self, *args, **kwargs):
+            return super().json(*args, **kwargs)
+
+        @classmethod
+        def model_json_schema(cls, *args, **kwargs):
+            return super(BaseModel, cls).schema_json(*args, **kwargs)
+
+        def model_copy(self, *args, **kwargs):
+            return super().copy(*args, **kwargs)
+
+else:
+    # TODO (ORQSDK-1025): remove the model base class and replace it with an alies to
+    # BaseModel
+    class BaseModel(pydantic.BaseModel):
+        """The pydantic BaseModel changed between V1 and V2.
+
+        As a result, workflow outputs generated prior to the V2 upgrade may not be
+        depickled correctly. The culpret is a change in behaviour of `__setstate__`.
+
+        This class adds a new `__setstate__` that wraps the V2 BaseModel `__setstate__`
+        and adds the missing behaviour back in.
+        """
+
+        def __setstate__(self, state: dict[Any, Any]) -> None:
+            state.setdefault("__pydantic_extra__", {})
+            state.setdefault("__pydantic_private__", {})
+
+            if "__pydantic_fields_set__" not in state:
+                state["__pydantic_fields_set__"] = state.get("__fields_set__")
+
+            super().__setstate__(state)
+
+
+class TypeAdapter:
+    """Accessor for Pydantic parsing.
+
+    If Pydantic V2 is installed, this class is a simple wrapper for
+    `pydantic.TypeAdapter`.
+
+    If Pydantic V1 is installed, this class acts as a translator between the V1-specific
+    `parse_X_as` methods and the V2 TypeAdapter style syntax we use in our code.
+    """
+
+    def __init__(self, model, *args, **kwargs):
+        if PYDANTICV1:
+            self._model = model
+        else:
+            self._typeadapter = pydantic.TypeAdapter(model, *args, **kwargs)
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏This can be generic because the underlying `TypeAdapter` is generic. Not sure if it's worth it.

```suggestion
_T = TypeVar("_T")

class TypeAdapter(Generic[_T]):
    """Accessor for Pydantic parsing.

    If Pydantic V2 is installed, this class is a simple wrapper for
    `pydantic.TypeAdapter`.

    If Pydantic V1 is installed, this class acts as a translator between the V1-specific
    `parse_X_as` methods and the V2 TypeAdapter style syntax we use in our code.
    """

    def __init__(self, model: Type[_T], *args, **kwargs):
        if PYDANTICV1:
            self._model = model
        else:
            self._typeadapter = pydantic.TypeAdapter[_T](model, *args, **kwargs)
```
and modify `typing` import to import `Generic`

Reply:
sounds good to me. I don't think it's a blocker either